sticker/src/dataset.rs

```diff
@@ -0,0 +1,284 @@
+use std::collections::HashMap;
+use std::io::{BufReader, Read, Seek, SeekFrom};
+
+use conllx::graph::Sentence;
+use conllx::io::{ReadSentence, Reader};
+use failure::Fallible;
+use ndarray::Array1;
+use rand::SeedableRng;
+use rand_xorshift::XorShiftRng;
+use sticker_encoders::SentenceEncoder;
+
+use crate::encoders::NamedEncoder;
+use crate::input::vectorizer::WordPieceVectorizer;
+use crate::input::{Tokenize, WordPieceTokenizer};
+use crate::tensor::{TensorBuilder, Tensors};
+use crate::util::RandomRemoveVec;
+
+/// A set of training/validation data.
+///
+/// A data set provides an iterator over the batches in that
+/// dataset.
+pub trait DataSet<'a> {
+    type Iter: Iterator<Item = Fallible<Tensors>>;
+
+    /// Get an iterator over the dataset batches.
+    ///
+    /// The sequence inputs are encoded with the given `vectorizer`,
+    /// the sequence labels using the `encoder`.
+    ///
+    /// Sentences longer than `max_len` are skipped. If you want to
+    /// include all sentences, you can use `usize::MAX` as the maximum
+    /// length.
+    fn batches(
+        self,
+        encoders: &'a [NamedEncoder],
+        tokenizer: &'a WordPieceTokenizer,
+        vectorizer: &'a WordPieceVectorizer,
+        batch_size: usize,
+        max_len: Option<usize>,
+        shuffle_buffer_size: Option<usize>,
+    ) -> Fallible<Self::Iter>;
+}
+
+/// A CoNLL-X data set.
+pub struct ConllxDataSet<R>(R);
+
+impl<R> ConllxDataSet<R> {
+    /// Construct a CoNLL-X dataset.
+    pub fn new(read: R) -> Self {
+        ConllxDataSet(read)
+    }
+
+    /// Returns an `Iterator` over `Result<Sentence, Error>`.
+    ///
+    /// Depending on the parameters the returned iterator filters
+    /// sentences by their lengths or returns the sentences in
+    /// sequence without filtering them.
+    ///
+    /// If `max_len` == `usize::MAX`, no filtering is performed.
```

Comment:
If `max_len` == None, right?

Reply:
Well, actually usize::MAX should have the same effect..

Reply:
True :). But updated.